File: social-media-post-service/src/main/java/com/fakechitor/socialmediapostservice/exception/GlobalExceptionHandler.java
package com.fakechitor.socialmediapostservice.exception;

import com.fakechitor.socialmediapostservice.dto.response.ExceptionMessageDto;
import io.jsonwebtoken.JwtException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.client.HttpServerErrorException;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;

@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(JwtException.class)
    ResponseEntity<ExceptionMessageDto> handleJwtException(JwtException e ){
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(new ExceptionMessageDto(e.getMessage()));
    }

    @ExceptionHandler(PostNotFoundException.class)
    ResponseEntity<ExceptionMessageDto> handlePostNotFoundException(PostNotFoundException e ){
        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new ExceptionMessageDto(e.getMessage()));
    }

    @ExceptionHandler(MethodArgumentTypeMismatchException.class)
    ResponseEntity<ExceptionMessageDto> handleMethodArgumentTypeMismatchException(MethodArgumentTypeMismatchException e){
        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ExceptionMessageDto(e.getMessage()));
    }

    @ExceptionHandler(ForbiddenAccessException.class)
    ResponseEntity<ExceptionMessageDto> handleForbiddenAccessException(ForbiddenAccessException e){
        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(new ExceptionMessageDto(e.getMessage()));
    }

    @ExceptionHandler(HttpServerErrorException.class)
    ResponseEntity<ExceptionMessageDto> handleHttpServerErrorException(HttpServerErrorException e){
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(new ExceptionMessageDto("Something went wrong =("));
    }
}

File: social-media-user-service/src/main/kotlin/com/fakechitor/socialmediauserservice/exception/GlobalExceptionHandler.kt
package com.fakechitor.socialmediauserservice.exception

import com.fakechitor.socialmediauserservice.dto.response.ExceptionMessageDto
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.client.HttpServerErrorException

@RestControllerAdvice
class GlobalExceptionHandler {
    @ExceptionHandler(UserNotFoundException::class)
    fun handleUserNotFoundException(e: UserNotFoundException) =
        ResponseEntity.status(HttpStatus.NOT_FOUND).body(ExceptionMessageDto(e.message))

    @ExceptionHandler(UserAlreadyExistsException::class)
    fun handleUserAlreadyExistsException(e: UserAlreadyExistsException) =
        ResponseEntity.status(HttpStatus.CONFLICT).body(ExceptionMessageDto(e.message))

    @ExceptionHandler(HttpServerErrorException::class)
    fun handleHttpServerErrorException(e: HttpServerErrorException?) =
        ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(ExceptionMessageDto("Something went wrong =("))

    @ExceptionHandler(SubscriptionNotFoundException::class)
    fun handleSubscriptionNotFoundException(e: SubscriptionNotFoundException) =
        ResponseEntity.status(HttpStatus.NOT_FOUND).body(ExceptionMessageDto(e.message))

    @ExceptionHandler(UserAlreadySubscribedException::class)
    fun handleUserAlreadySubscribedException(e: UserAlreadySubscribedException) =
        ResponseEntity.status(HttpStatus.CONFLICT).body(ExceptionMessageDto(e.message))

    @ExceptionHandler(UserNotFriendException::class)
    fun handleUserNotFriendException(e: UserNotFriendException) =
        ResponseEntity.status(HttpStatus.FORBIDDEN).body(ExceptionMessageDto(e.message))
}
